api/cloud/oc_cloud.c

```diff
@@ -106,48 +115,119 @@ cloud_close_endpoint(oc_endpoint_t *cloud_ep)
     oc_connectivity_end_session(cloud_ep);
 #endif /* OC_TCP */
   }
 }
 
 #ifdef OC_SECURITY
 static void
-cloud_deregister_on_reset_internal(oc_cloud_context_t *ctx,
-                                   oc_cloud_status_t status, void *data)
+cloud_deregister_on_reset_handler(oc_cloud_context_t *ctx,
+                                  oc_cloud_status_t status, void *data)
 {
   (void)status;
   (void)data;
+  OC_DBG("[Cloud] cloud_deregister_on_reset_handler device=%zu", ctx->device);
   cloud_context_clear(ctx);
-  OC_DBG("[Cloud] cloud_deregister_on_reset_internal device=%zu", ctx->device);
 }
+
+static oc_event_callback_retval_t
+cloud_deregister_context_clear_async(void *data)
+{
+  oc_cloud_context_t *ctx = (oc_cloud_context_t *)data;
+  cloud_context_clear(ctx);
+  return OC_EVENT_DONE;
+}
+
+static void
+cloud_deregister_on_reset_async_handler(oc_cloud_context_t *ctx,
+                                        oc_cloud_status_t status, void *data)
+{
+  (void)status;
+  (void)data;
+  OC_DBG("[Cloud] cloud_deregister_on_reset_async_handler device=%zu",
+         ctx->device);
+  // this call might be invoked because of timeout (delayed call of
+  // oc_ri_remove_client_cb_with_notify_503 when the request is fired), however
+  // cloud_context_clear among other things closes the connected endpoint. This
+  // also invokes oc_ri_remove_client_cb_with_notify_503 for this call,
+  // causing memory issues. We schedule the context clear in a delayed callback,
+  // so this call removes itself from the queue of calls for the endpoint.
+  cloud_reset_delayed_callback(ctx, cloud_deregister_context_clear_async, 0);
+}
+
+static int
+cloud_deregister_on_reset(oc_cloud_context_t *ctx, bool sync, uint16_t timeout)
+{
+  OC_DBG("[Cloud] cloud deregister on reset");
+  oc_cloud_cb_t handler = sync ? &cloud_deregister_on_reset_handler
+                               : &cloud_deregister_on_reset_async_handler;
+  int err = cloud_deregister(ctx, sync, timeout, handler, ctx);
+  if (err == 0 || err == CLOUD_DEREGISTER_ERROR_ALREADY_DEREGISTERING) {
+    return 0;
+  }
+  return -1;
+}
+
+static oc_event_callback_retval_t
+cloud_deregister_on_reset_async(void *data)
+{
+  cloud_api_param_t *p = (cloud_api_param_t *)data;
+  oc_cloud_context_t *ctx = p->ctx;
+  uint16_t timeout = p->timeout;
+  free_api_param(p);
+  if (cloud_deregister_on_reset(ctx, /*sync*/ false, timeout) != 0) {
+    OC_ERR("[Cloud] failed to deregister from cloud");
+    cloud_context_clear(ctx);
+  }
+  return OC_EVENT_DONE;
+}
+
+bool
+cloud_reset_execute(oc_cloud_context_t *ctx, bool sync, uint16_t timeout)
+{
+  cloud_manager_stop(ctx);
+  if (sync) {
+    return cloud_deregister_on_reset(ctx, true, 0) == 0;
+  }
+
+  cloud_api_param_t *p = alloc_api_param();
+  if (p == NULL) {
+    OC_ERR("[Cloud] cannot allocate cloud parameters for reset");
+    return false;
+  }
+  p->ctx = ctx;
+  p->timeout = timeout;
+  cloud_reset_delayed_callback(p, cloud_deregister_on_reset_async, 0);
```

Comment:
this doesn't remove `cloud_deregister_on_reset_async` from delayed callbacks -> it always add new one.